### PR TITLE
Feat: quickloot in cpp

### LINKED ISSUE #1030: Quickloot issues:
Quickloot has a couple issues that I've found thus far:

First issue:

If you select a brown backpack inside of your main as the container for "others" category, and you have different backpacks inside of your brown backpack, once the first is filled up, you won't be able to pick up loots, until you put a identical backpack inside of it

ex:

1- select a brown backpack as your 

image

2- put any other type of backpack inside (beach backpack for example)

image

3- fill it up with loot

image

4- try to pick up more loot

image
image

5- substitute the beach backpack with another brown backpack and try again

image
image
image

And as you can see, CAP is not the problem:

image

Second issue:

Another issue related to the last one, as you can see through those images, if your container is filled up, the countable drops will still be sent to the other backpack, instead of keep stacking on the first and so backpacks, even though their 100x limit has not been reached yet.

### LINKED ISSUE #1137: Monster still fast, quickloot problems.
sorry for write again.
using the last build. and i still ultra-fast monsters? can help me to solve it?
thanks!!!
another problem.
recent corpses can'r move them. i only try in rats xD
last problem.
quickloot isn't work, console say this in the server start

> [Warning - Items::parseItemNode] Unknown key value: quicklootcategory

in-game says the following

>  Lua Script Error: [Modules Interface]
> 
> data/modules/scripts/quickloot/quickloot.lua:onRecvbyte
> data/modules/scripts/quickloot/quickloot.lua:131: attempt to call method 'canOpenCorpse' (a nil value)
> stack traceback:
>         [C]: in function 'canOpenCorpse'
>         data/modules/scripts/quickloot/quickloot.lua:131: in function <data/modules/scripts/quickloot/quickloot.lua:74>
> 

the last error says the following. I do not know what it is.

> Lua Script Error: [Event Interface]
> data/events/scripts/player.lua:Player@onGainExperience
> data/global.lua:133: bad argument #1 to 'ipairs' (table expected, got nil)
> stack traceback:
>         [C]: at 0x7ff74807a6d0
>         [C]: in function 'ipairs'
>         data/global.lua:133: in function 'getRateFromTable'
>         data/events/scripts/player.lua:764: in function <data/events/scripts/player.lua:751>
> 
> Lua Script Error: [Event Interface]
> data/events/scripts/monster.lua:Monster@onDropLoot
> data/events/scripts/monster.lua:31: attempt to call method 'updateKillTracker' (a nil value)
> stack traceback:
>         [C]: in function 'updateKillTracker'
>         data/events/scripts/monster.lua:31: in function <data/events/scripts/monster.lua:1>
> 
regards!

### PR DESCRIPTION
**Pull request by dudantas**
Removed quickloot in Lua and added in CPP

The system is not mine, I don't know the authorship, it was obtained through a donation from a base. If the owner wants to claim the credits, leave a message that I will gladly post.

Please test it and report if you found any issue.

### CODE REVIEW

File: src/enums.h [hunk 1]
```diff
@@ -94,6 +94,8 @@ enum itemAttrTypes : uint32_t {
 	ITEM_ATTRIBUTE_SPECIAL = 1 << 23,
 	ITEM_ATTRIBUTE_IMBUINGSLOTS = 1 << 24,
 	ITEM_ATTRIBUTE_OPENCONTAINER = 1 << 25,
+	ITEM_ATTRIBUTE_QUICKLOOTCONTAINER = 1 << 25,
```
Comment (vinioliveirasilva) [inline src/enums.h]: this enum should have the same value as ITEM_ATTRIBUTE_OPENCONTAINER?

File: src/protocolgame.cpp [hunk 2]
```diff
@@ -458,6 +504,10 @@ void ProtocolGame::parsePacket(NetworkMessage& msg)
 		case 0x8C: parseLookAt(msg); break;
 		case 0x8D: parseLookInBattleList(msg); break;
 		case 0x8E: /* join aggression */ break;
+		case 0x8F: parseQuickLoot(msg); break;
+		case 0x90: parseLootContainer(msg); break;
+		case 0x91: parseQuickLootBlackWhitelist(msg); break;
+		case 0x92: parseResquestLockItems(); break;
```
Comment (vinioliveirasilva) [inline src/protocolgame.cpp]: typo error

File: src/protocolgame.cpp [hunk 3]
```diff
@@ -903,6 +953,52 @@ void ProtocolGame::parseLookInBattleList(NetworkMessage& msg)
...
+	listedItems.reserve(size);
+
+	for (int i = 0; i < size; i++) {
+		listedItems.push_back(msg.get<uint16_t>());
+	}
+
+	addGameTask(&Game::playerQuickLootBlackWhitelist, player->getID(), filter, listedItems);
+}
+
+void ProtocolGame::parseResquestLockItems()
```
Comment (vinioliveirasilva) [inline src/protocolgame.cpp]: typo error

File: src/protocolgame.h [hunk 4]
```diff
@@ -104,6 +106,12 @@ class ProtocolGame final : public Protocol
 		void parseSay(NetworkMessage& msg);
 		void parseLookAt(NetworkMessage& msg);
 		void parseLookInBattleList(NetworkMessage& msg);
+
+		void parseQuickLoot(NetworkMessage& msg);
+		void parseLootContainer(NetworkMessage& msg);
+		void parseQuickLootBlackWhitelist(NetworkMessage& msg);
+		void parseResquestLockItems();
```
Comment (vinioliveirasilva) [inline src/protocolgame.h]: typo error

**Comment by vinioliveirasilva** [comment]
Really good job, i going to test.

**Comment by joseluis2g** [comment]
why are those opened attributes added?

**Comment by dudantas** [comment]
> why are those opened attributes added?

I added it by mistake, not necessary in quickloot

**Comment by dudantas** [comment]
> Really good job, i going to test.

Fixed the errors, thank you

File: src/game.cpp [hunk 5]
```diff
@@ -1786,6 +1799,295 @@ ReturnValue Game::internalTeleport(Thing* thing, const Position& newPos, bool pu
 	return RETURNVALUE_NOTPOSSIBLE;
 }
 
+void Game::internalQuickLootCorpse(Player* player, Container* corpse)
+{
+	std::vector<Item*> itemList;
```
Comment (lgrossi) [inline src/game.cpp:1833]: Please validate if player Nd corpse are not null

File: src/game.cpp [hunk 6]
```diff
@@ -1786,6 +1799,295 @@ ReturnValue Game::internalTeleport(Thing* thing, const Position& newPos, bool pu
...
+	} else {
+		player->sendTextMessage(MESSAGE_EVENT_DEFAULT, ss.str());
+	}
+
+	player->lastQuickLootNotification = OTSYS_TIME();
+}
+
+ReturnValue Game::internalQuickLootItem(Player* player, Item* item, ObjectCategory_t category /* = OBJECTCATEGORY_DEFAULT*/)
+{
+	bool fallbackConsumed = false;
```
Comment (lgrossi) [inline src/game.cpp:1974]: same here, be safe about the pointers parameters.

File: src/game.cpp [hunk 7]
```diff
@@ -1786,6 +1799,295 @@ ReturnValue Game::internalTeleport(Thing* thing, const Position& newPos, bool pu
...
+			lootContainer = fallbackItem->getContainer();
+			it = lootContainer->iterator();
+
+			fallbackConsumed = true;
+		}
+	} while (remainderCount != 0);
+	return ret;
+}
+
+ObjectCategory_t Game::getObjectCategory(const Item* item)
```
Comment (lgrossi) [inline src/game.cpp:2068]: Same

File: src/game.cpp [hunk 8]
```diff
@@ -3244,6 +3546,243 @@ void Game::playerLookInBattleList(uint32_t playerId, uint32_t creatureId)
 	g_events->eventPlayerOnLookInBattleList(player, creature, lookDistance);
 }
 
+void Game::playerQuickLoot(uint32_t playerId, const Position& pos, uint16_t spriteId, uint8_t stackPos, Item* defaultItem)
```
Comment (lgrossi) [inline src/game.cpp:3601]: same here

File: src/player.cpp [hunk 9]
```diff
@@ -791,6 +795,90 @@ void Player::onReceiveMail() const
...
+
+	if (category != OBJECTCATEGORY_DEFAULT) {
+		// firstly, fallback to default
+		return getLootContainer(OBJECTCATEGORY_DEFAULT);
+	}
+
+	return nullptr;
+}
+
+void Player::checkLootContainers(const Item* item)
```
Comment (lgrossi) [inline src/player.cpp:854]: same

File: src/items.cpp [hunk 10]
```diff
@@ -42,6 +42,45 @@ void Items::clear()
 	nameToItems.clear();
 }
 
+using LootTypeNames = std::unordered_map<std::string, ItemTypes_t>;
+
+LootTypeNames lootTypeNames = {
```
Comment (lgrossi) [inline src/items.cpp:47]: really nice this one! great job

File: src/player.h [hunk 11]
```diff
@@ -1372,6 +1392,18 @@ class Player final : public Creature, public Cylinder
 			lastMarketInteraction = OTSYS_TIME();
 		}
 
+		bool isQuickLootListedItem(const Item* item) const {
+			auto it = std::find(quickLootListClientIds.begin(), quickLootListClientIds.end(), item->getClientID());
```
Comment (lgrossi) [inline src/player.h:1426]: same here, what if item is null?

File: src/protocolgame.cpp [hunk 12]
```diff
@@ -48,6 +48,61 @@ extern Modules* g_modules;
...
+	} else if (it.isContainer() && player->getOperatingSystem() <= CLIENTOS_NEW_WINDOWS) {
+		msg.addByte(0x00);
+	}
+
+	if (it.isAnimation) {
+		msg.addByte(0xFE); // random phase (0xFF for async)
+	}
+}
+
+void ProtocolGame::AddItem(NetworkMessage& msg, const Item* item)
```
Comment (lgrossi) [inline src/protocolgame.cpp:70]: same here, what if item is null

File: src/protocolgame.cpp [hunk 13]
```diff
@@ -1745,6 +1841,36 @@ void ProtocolGame::sendContainer(uint8_t cid, const Container* container, bool h
...
+	for (auto it : quickLoot) {
+		msg.addByte(it.first);
+		msg.add<uint16_t>(it.second->getClientID());
+	}
+
+	writeToOutputBuffer(msg);
+}
+
+void ProtocolGame::sendLootStats(Item* item)
+{
```
Comment (lgrossi) [inline src/protocolgame.cpp:1905]: same here, what if item is null?

**Review by lgrossi — DISMISSED** [review]
Great job, it looks good in general. However, I've pointed out a lot of unsafe pointer access, we should fix it.

File: src/protocolgame.cpp [hunk 14]
```diff
@@ -903,6 +953,52 @@ void ProtocolGame::parseLookInBattleList(NetworkMessage& msg)
...
+	if (action == 0) {
+		ObjectCategory_t category = (ObjectCategory_t)msg.getByte();
+		Position pos = msg.getPosition();
+		uint16_t spriteId = msg.get<uint16_t>();
+		uint8_t stackpos = msg.getByte();
+		addGameTask(&Game::playerSetLootContainer, player->getID(), category, pos, spriteId, stackpos);
+	} else if (action == 1) {
+		ObjectCategory_t category = (ObjectCategory_t)msg.getByte();
+		addGameTask(&Game::playerClearLootContainer, player->getID(), category);
+	} else if (action == 3) {
```
Comment (gpedro) [inline src/protocolgame.cpp]: missing action 2
- open category container

Comment (lgrossi) [reply src/protocolgame.cpp]: Maybe the actions could be using an enum cast instead?

Comment (joseluis2g) [reply src/protocolgame.cpp]: I would also prefer some uint8 enum if there is more than 2 options.

**Comment by joao-blip** [comment]
I noticed that when you log out the character is not saving the backpacks.
and if you kill a lot of mobs when you take out the dead monsters to get the ones below the auto loot doesn't work

**Comment by joao-blip** [comment]
any news on this subject?
so I tested the only bugs I found were the two mentioned above.
rest is working

**Comment by MajestyOTBR** [comment]
@gpedro @lgrossi PR waiting for your review.

**Comment by gpedro** [comment]
@joao-blip can you test again?

--

If you like our project, feel free to pay us a coffee, pizza or something else. Click here check donations methods: https://opentibiabr.github.io/donation

**Comment by gpedro** [comment]
know issue:
- creature products are not recognized.

**Comment by amoxicilinaaa** [comment]
erro console ao compilar, servidor otservbr-global-quickloot-cpp

```
1>C:\Users\Lucas\Desktop\otservbr-global-quickloot-cpp\src\housetile.h(50,1): warning C4266: 'void Tile::internalAddThing(Thing *)': nenhuma substituição disponível para função membro virtual da base 'Tile'; função é ocultada (compilando arquivo de origem ..\src\protocollogin.cpp)
1>C:\Users\Lucas\Desktop\otservbr-global-quickloot-cpp\src\tile.h(270): message : consulte a declaração de 'Tile::internalAddThing' (compilando arquivo de origem ..\src\protocollogin.cpp)
1>C:\Users\Lucas\Desktop\otservbr-global-quickloot-cpp\src\tile.h(153): message : consulte a declaração de 'Tile' (compilando arquivo de origem ..\src\protocollogin.cpp)
1>protocolstatus.cpp
```
Salvando as configurações após relogar, mais agora os itens não vão para as backpacks especificadas e alguns nem são loteados para nenhuma backpack.

**Comment by MajestyOTBR** [comment]
São warnings, não estão relacionados ao quickloot.
Especifique melhor o que está acontecendo, teste cada opção do quickloot, configure as backpacks nas opções, adicione os itens a lista, verifique no ciclopedia.

**Comment by amoxicilinaaa** [comment]
erro console ao compilar, servidor otservbr-global-quickloot-cpp

```
1>C:\Users\Lucas\Desktop\otservbr-global-quickloot-cpp\src\housetile.h(50,1): warning C4266: 'void Tile::internalAddThing(Thing *)': nenhuma substituição disponível para função membro virtual da base 'Tile'; função é ocultada (compilando arquivo de origem ..\src\protocollogin.cpp)
1>C:\Users\Lucas\Desktop\otservbr-global-quickloot-cpp\src\tile.h(270): message : consulte a declaração de 'Tile::internalAddThing' (compilando arquivo de origem ..\src\protocollogin.cpp)
1>C:\Users\Lucas\Desktop\otservbr-global-quickloot-cpp\src\tile.h(153): message : consulte a declaração de 'Tile' (compilando arquivo de origem ..\src\protocollogin.cpp)
1>protocolstatus.cpp
```
Salvando as configurações após relogar, mais agora os itens não vão para as backpacks especificadas e alguns nem são loteados para nenhuma backpack.

**Comment by dudantas** [comment]
> Alguns itens não vão para as backpacks configuradas e os foods não são loteados.
> 
> Some items do not go as configured backpacks and the food is not blended.

Pode testar novamente? Subi algumas correções.

**Comment by dudantas** [comment]
> know issue:
> 
> * creature products are not recognized.

Fixed

**Comment by Attack787** [comment]
valuables is wrong. When you pick jewelry, him taks to main backpack and not to backpack configured.

**Comment by lekzinn** [comment]
- When i have a BP of another type inside main container, he doesn't recognize and use as fallback backpack (if enable)
- When the backpack is full, stackable items are not grouped correctly. Even though the container is full, they should group with their stackable items before going to another backpack
- Conflicting files

See screenshot: 

image

**Comment by marlon36** [comment]
hi excuse me if I write where I shouldn't, I would like to test the autoloot system but I don't know how to activate I have downloaded your branch of development and the last compilation 1.0.1128 but I can't make it work, is it enabled or do I have to compile something?

**Comment by MajestyOTBR** [comment]
The correct executable for quickloot on AppVeyor is 1.0.1121 that correspond to the last commit did here ([2018eba](**https://github.com/opentibiabr/otservbr-global/pull/1400/commits/2018ebafc62ef34697a5cfca9a47766fce09fdf8**)):
https://ci.appveyor.com/project/opentibiabr/otservbr-global/builds/34191541

**Comment by MajestyOTBR** [comment]
The correct executable for quickloot on AppVeyor is 1.0.1121 that correspond to the last commit did here (2018eba):
https://ci.appveyor.com/project/opentibiabr/otservbr-global/builds/34191541

And you need to change the files:
https://github.com/opentibiabr/otservbr-global/pull/1400/files

**Comment by marcosvf132** [comment]
Script is not checking the childs backpacks, only the first one. (Missing 'return' on script)

**Review by majestyotbr — APPROVED** [review]

**Comment by majestyotbr** [comment]
Tested, working. I didn't find any issue.

**Review by YohanAugusto — APPROVED** [review]

**Comment by thatmichaelguy** [comment]
A migration for this PR would be nice
The "Use main container as fallback" tickbox doesn't save
Loot will go to the main backpack and then any other backpack as each gets filled regardless of the allocation - not sure what happens on real tibia

Other than that, seems to be working nicely

**Pull request merged**